File: jserver-core/src/main/java/com/github/gr3gdev/jserver/socket/SocketReader.java
package com.github.gr3gdev.jserver.socket;

import com.github.gr3gdev.jserver.http.Request;
import com.github.gr3gdev.jserver.http.impl.RequestImpl;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.Set;

/**
 * SocketReader.
 *
 * @author Gregory Tardivel
 */
public class SocketReader implements Runnable {

    private final Socket socket;
    private final Set<SocketEvent> socketEvents;

    public SocketReader(Socket socket, Set<SocketEvent> socketEvents) {
        this.socket = socket;
        this.socketEvents = socketEvents;
    }

    @Override
    public void run() {
        // Request
        try (final InputStream inputStream = socket.getInputStream();
             final OutputStream outputStream = socket.getOutputStream()) {
            final Request request = new RequestImpl(socket.getRemoteSocketAddress().toString(), inputStream);
            // Search event
            socketEvents.stream()
                    .filter(e -> e.match(request))
                    .forEach(e -> e.getRouteListener().handleEvent(request, outputStream));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: jserver-core/src/main/kotlin/com/github/gr3gdev/jserver/socket/SocketReader.kt
package com.github.gr3gdev.jserver.socket

import com.github.gr3gdev.jserver.http.impl.RequestImpl
import java.net.Socket

/**
 * SocketReader.
 *
 * @author Gregory Tardivel
 */
internal class SocketReader(private val socket: Socket, private val socketEvents: List<SocketEvent>) : Runnable {

    override fun run() {
        socket.use {
            // Request
            val request = RequestImpl(it.remoteSocketAddress.toString(), it.getInputStream())
            // Search event
            socketEvents.filter { e -> e.match(request) }
                    .forEach { e -> e.routeListener.handleEvent(request, it.getOutputStream()) }
        }
    }

}
